src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,1006 @@
+                                extensionv[2 * i] = strtup->name;
+                                extensionv[2 * i + 1] = strtup->value;
+                        }
+                        rd_kafka_oauthbearer_set_token(rk, jws,
+                                now_wallclock_millis + life_seconds * 1000,
+                                principal, extensionv, 2 * extension_pair_count,
+                                errstr, sizeof(errstr));
+                        rd_free(jws);
+                        rd_free(extensionv);
+                }
+        }
+        rd_free(principal_claim_name);
+        rd_free(principal);
+        rd_free(scope_claim_name);
+        rd_free(scope_csv_text);
+        rd_list_destroy(&scope);
+        rd_list_destroy(&extensions);
+}
+
+/**
+ * @brief Close and free authentication state
+ */
+static void rd_kafka_sasl_oauthbearer_close (rd_kafka_transport_t *rktrans) {
+        struct rd_kafka_sasl_oauthbearer_state *state =
+                rktrans->rktrans_sasl.state;
+
+        if (!state)
+                return;
+
+        RD_IF_FREE(state->server_error_msg.ptr, rd_free);
+        rd_free(state->token_value);
+        rd_free(state->md_principal_name);
+        rd_list_destroy(&state->extensions);
+        rd_free(state);
+}
+
+
+
+/**
+ * @brief Build client-first-message
+ */
+static void
+rd_kafka_sasl_oauthbearer_build_client_first_message (
+        rd_kafka_transport_t *rktrans,
+        rd_chariov_t *out) {
+        struct rd_kafka_sasl_oauthbearer_state *state =
+                rktrans->rktrans_sasl.state;
+
+        /*
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * kvsep          = %x01
+         * key            = 1*(ALPHA)
+         * value          = *(VCHAR / SP / HTAB / CR / LF )
+         * kvpair         = key "=" value kvsep
+         * ;;gs2-header     = See RFC 5801
+         * client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+         */
+
+        static const char *gs2_header = "n,,";
+        static const char *kvsep = "\x01";
+        const int kvsep_size = strlen(kvsep);
+        int extension_size = 0;
+        int i;
+        char *buf;
+        int size_written;
+
+        for (i = 0 ; i < rd_list_cnt(&state->extensions) ; i++) {
+                rd_strtup_t *extension = rd_list_elem(&state->extensions, i);
+                // kvpair         = key "=" value kvsep
+                extension_size += strlen(extension->name) + 1 // "="
+                        + strlen(extension->value) + kvsep_size;
+        }
+
+        // client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+        out->size = strlen(gs2_header) + kvsep_size
+                + strlen("auth=Bearer ") + strlen(state->token_value)
+                + kvsep_size + extension_size + kvsep_size;
+        out->ptr = rd_malloc(out->size+1);
+
+        buf = out->ptr;
+        size_written = 0;
+        size_written += rd_snprintf(buf, out->size+1 - size_written,
```

Comment:
do note that snprintf() will return the *needed* size if the string would not fit, which should not happen here but just to make sure, do this:
```
r = rd_snprintf(buf, out->size+1 - size_written, ..);
rd_assert(r <= out->size+1 - size_written);
r += size_written;
```

Reply:
👍 Though I used the assert line `rd_assert(r < out->size+1 - size_written)` as per https://linux.die.net/man/3/snprintf:

> a return value of size or more means that the output was truncated